EmployeeList/ViewModels/CustomImageViewModel.swift

```diff
@@ -0,0 +1,142 @@
+                return url
+            }
+            let endIndex = imageURLComponents.endIndex
+            let id = imageURLComponents[endIndex - 2]
+            let size = imageURLComponents[endIndex - 1]
+            return id + size
+        }
+        return url
+    }
+    
+    func bindInputsToOutputs() {
+        let imageKey = buildImageKey(self.imageURL)
+
+        if self.imageService.existsInDisk(imageKey) {
+            print("Load image from disk")
+            loadFromDisk(from: imageKey)
+            loadImage()
+        } else {
+            print("Load image from web")
+            loadFromWeb(from: self.imageURL)
+            loadImage()
+        }
+
+    }
+    
+    
+    // MARK: Bind inputs
+    func loadFromDisk(from key: String) {
+
+        let publisher = onAppearSubject.flatMap { _ ->  AnyPublisher<Data, Error> in
+            return self.imageService.loadFromDisk(forKey: key)
+        }
+        let diskStream = publisher.subscribe(imageSubject)
+        
+        cancellables["disk"] = diskStream
+    }
+    
+    func loadFromWeb(from url: String) {
+        
+        let endpoint = ImageRequest(path: url)
+        let publisher = onAppearSubject.flatMap { _ -> AnyPublisher<Data, Error> in
+            return self.imageService.call(from: endpoint)
+        }
+        let webStream = publisher.subscribe(imageSubject)
+        cancellables["web"] = webStream
+    }
+    
+    // MARK: Only save the image to disk the first time the image is downloaded from the web
+    func saveToDisk(_ image: UIImage, _ key: String) {
+
+        let publisher = diskSaveSubject.flatMap { _ -> AnyPublisher<Bool, Error> in
+            return self.imageService.saveToDisk(self.image, forKey: key)
+        }
+        
+        let diskSaveStream = publisher
+            .map { $0 }
+            .sink(receiveCompletion: { result in
+                switch result {
+                case .failure(let error): do {
+                    self.isErrorShown = true
+                    self.errorMessage = error.localizedDescription
+                    }
+                case .finished:
+                    break
+                }
+            }, receiveValue: { data in
+                print("Successfully stored image, boolean is \(data)")
+            })
+        
+        diskSaveSubject.send()
+        cancellables["diskSave"] = diskSaveStream
+    }
+    
+    // MARK: Bind output
+    func loadImage() {
+        
+        let imageStream = imageSubject
+            .map{ $0 }
+            .sink(receiveCompletion: { result in
+                switch result {
+                case .failure(let error): do {
+                    self.isErrorShown = true
+                    self.errorMessage = error.localizedDescription
+                    }
+                case .finished:
+                    break
+                }
+            }, receiveValue: { data in
+                self.image = UIImage(data: data) ?? UIImage(named: "iu")!
```

Comment:
not a great image name

Reply:
I should rename "iu.png" to "defaultAvatar.png", right?

Reply:
that's better.